### PR TITLE
[WIP] engine: Add fd cache

### PR DESCRIPTION
**Pull request by uran0sH**

### CODE REVIEW

File: src/server/storage_engine/default_engine.rs [hunk 1]
```diff
@@ -180,21 +200,27 @@ impl StorageEngine for DefaultEngine {
         }
 
         let local_file_name = generate_local_file_name(&self.root, &path);
-        let oflag = OFlag::O_RDONLY;
+        let oflag = OFlag::O_WRONLY;
```
Comment (luanshaotong) [inline src/server/storage_engine/default_engine.rs:203]: What will happen if write after reading? Will it return "read only fd"?

Comment (uran0sH) [reply src/server/storage_engine/default_engine.rs]: It's a mistake.

File: src/common/cache.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                (*(next_node_locked.0.unwrap().as_ptr())).prev.lock().0 = prev_node_locked.0;
+            }
+            node.into_val()
+        }
+    }
+
+    pub fn reinsert_front(&self, key: &[u8], node: NonNull<Node<T>>) {
+        let data = self.remove(node);
+        self.insert_front(key, data);
+    }
```
Comment (mond77) [inline src/common/cache.rs:116]: it is better to check if the node was the head here.

**Comment by HHoflittlefish777** [comment]
Using aysnc flush is better.

**Comment by uran0sH** [comment]
before:
Screen Shot 2023-02-15 at 15 37 25
after:
Screen Shot 2023-02-20 at 17 38 24

**Pull request closed without merging**